common/src/main/java/com/wynntils/gui/render/MapRenderer.java

```diff
@@ -99,10 +116,235 @@ public static void renderMapQuad(
+            boolean followPlayerRotation,
+            float mapTopX,
+            float mapLeftZ,
+            float mapBottomX,
+            float mapRightZ) {
+        List<ServicePoi> servicePois = MapModel.getAllPois().stream()
+                .filter(poi -> {
+                    if (!(poi instanceof ServicePoi)) return false;
+
+                    return isPoiVisible(
+                            mapCenterX,
+                            mapCenterZ,
+                            centerX,
+                            centerZ,
+                            scale,
+                            mapTopX,
+                            mapLeftZ,
+                            mapBottomX,
+                            mapRightZ,
+                            poi);
+                })
+                .map(poi -> (ServicePoi) poi)
+                .toList();
+
+        if (mouseCoordinates != null) {
+            hovered = null;
+        }
+
+        final double rotationRadians = Math.toRadians(McUtils.player().getYRot());
+        final float sinRotationRadians = (float) StrictMath.sin(rotationRadians);
+        final float cosRotationRadians = (float) -StrictMath.cos(rotationRadians);
+
+        for (ServicePoi servicePoi : servicePois) {
+            renderServicePoi(
+                    poseStack,
+                    mouseCoordinates,
+                    mapCenterX,
+                    mapCenterZ,
+                    centerX,
+                    centerZ,
+                    followPlayerRotation,
+                    scale,
+                    poiScale,
+                    sinRotationRadians,
+                    cosRotationRadians,
+                    servicePoi);
+        }
+
+        if (McUtils.mc().screen instanceof MainMapScreen mainMapScreen) {
+            mainMapScreen.setHovered(hovered);
+        }
+
+        if (hovered != null && mouseCoordinates != null) {
+            float textureXPosition = getRenderX(hovered, mapCenterX, centerX, 1f / scale);
+            float textureZPosition = getRenderZ(hovered, mapCenterZ, centerZ, 1f / scale);
+
+            float width = hovered.getIcon().width() * poiScale;
+            float height = hovered.getIcon().height() * poiScale;
+
+            poseStack.pushPose();
+
+            float renderX = textureXPosition - width / 2f;
+            float renderZ = textureZPosition - height / 2f;
+            poseStack.translate(renderX, renderZ, 0);
+
+            FontRenderer.getInstance()
+                    .renderText(
+                            poseStack,
+                            hovered.getName(),
+                            width / 2f,
+                            20,
+                            CommonColors.GREEN,
+                            HorizontalAlignment.Center,
+                            VerticalAlignment.Top,
+                            FontRenderer.TextShadow.OUTLINE);
+
+            poseStack.popPose();
+        }
+    }
+
+    private static boolean isPoiVisible(
```

Comment:
This should be inlined

Reply:
Magicus wanted this extracted

Reply:
Actually I requested that he extracted it! 😆 For clarity.